### PR TITLE
Cache some properties on Dataset

### PR DESCRIPTION
**Pull request by takluyver**
In #1391, we noticed that repeated small reads from the same dataset are surprisingly slow. Profiling showed that quite a bit of time was wasted re-fetching and calculating things which don't change from read to read.

First, the headline figures:

- Simple `ds[0:10]` reads are about 2x faster (1.2 s -> 0.55 s)
- `read_direct()` about 25% faster (0.98 s -> 0.75 s)

Times are for 10k reads of 10 numbers each from a 1D integer dataset. [Details in this comment](https://github.com/h5py/h5py/issues/1391#issuecomment-542080099), but the timings given there are done with a different method.

The fastest I've got with Python code calling the low-level API directly is 0.22 s.

-----

For properties which can't change (like whether a dataset is empty), I've used a standard [@cached_property](https://github.com/pydanny/cached-property) decorator. They'll be calculated once on first access, then saved.

Dataset shape and size can change, either if the file is writeable (through `H5Dset_extent`) or if it's being read in SWMR mode (when `H5Drefresh` is called). I've taken a compromise position: 

* For SWMR mode the `ds.refresh()` call also clears the cache. This assumes that people keeping a `Dataset` object around will use its refresh method, not the low-level API.
* If the file is writeable, it doesn't use the cache at all, so it can't get outdated if you modify it with the low-level API.

The code could be simpler if we cached shape & size for writeable files as well - but it would mean that after manipulating a dataset with the low-level API you need to call `.refresh()` or get a new high-level Dataset object.

### CODE REVIEW

**Comment by takluyver** [comment]
Another tweak: using a simple Python function in place of `numpy.product()` in `SimpleSelection.broadcast()`.

- `ds[0:10]` reading unchanged
- `read_direct()` gets about 30% faster again (0.75 s -> 0.52 s). This is almost a 2x speedup from before this PR.

-----

With small amounts of data, `numpy.product()` is about an order of magnitude slower than the hand-coded `product()` function I added. E.g. with a tuple of 2 numbers, numpy takes ~8 µs, and the custom function about 200 ns.

Of course, this only works for small data, and possibly only for data that's not already in an array. But in various places we have one number per dimension, like this, and [HDF5 only allows up to 32 dimensions](https://support.hdfgroup.org/HDF5/faq/limits.html). Even if that limit were removed, I suspect there are very few use cases which need more dimensions.

File: h5py/_hl/dataset.py [hunk 1]
```diff
@@ -384,8 +404,17 @@ def fillvalue(self):
...
-    def __init__(self, bind):
+    def _extent_type(self):
+        return self.id.get_space().get_simple_extent_type()
+
+    @cached_property
+    def _is_empty(self):
+        return self._extent_type == h5s.NULL
+
+    @with_phil
+    def __init__(self, bind, readonly=False):
```
Comment (tacaswell) [inline h5py/_hl/dataset.py]: make `readonly` kwarg only?

File: setup.py [hunk 2]
```diff
@@ -32,7 +32,7 @@
 NUMPY_DEP = 'numpy>=1.7'
 
 # these are required to use h5py
-RUN_REQUIRES = [NUMPY_DEP,]
+RUN_REQUIRES = [NUMPY_DEP, "cached-property"]
```
Comment (tacaswell) [inline setup.py]: Does this interact nicely with py3.8?

File: h5py/_hl/base.py [hunk 3]
```diff
@@ -418,3 +418,15 @@ def __eq__(self, other):
 
     def __repr__(self):
         return "Empty(dtype={0!r})".format(self.dtype)
+
+
+def product(nums):
```
Comment (tacaswell) [inline h5py/_hl/base.py]: do we want to make this `_product` to protect it a bit from users?

File: h5py/_hl/dataset.py (hunk 1 above)
Comment (takluyver) [reply h5py/_hl/dataset.py]: Good idea, done.

File: setup.py (hunk 2 above)
Comment (takluyver) [reply setup.py]: No idea! Is there something you're aware of that might break it?

File: h5py/_hl/base.py (hunk 3 above)
Comment (takluyver) [reply h5py/_hl/base.py]: I figure that anything you have to import from `h5py._hl` is effectively private anyway - the public API is exposed on the top-level `h5py` module. But I'm happy to add a `_` if you prefer.

**Comment by takluyver** [comment]
I'll merge this in another day or two unless anyone objects or wants more time to review it.

**Comment by takluyver** [comment]
Merging on the strength of @tacaswell's approval and lack of objections. There's still time to make changes before release.

**Pull request merged**